app/src/main/cpp/skyline/soc/gm20b/engines/inline2memory.cpp

```diff
@@ -19,13 +20,50 @@ namespace skyline::soc::gm20b::engine {
         if (state.launchDma.completion == RegisterState::DmaCompletionType::ReleaseSemaphore)
             throw exception("Semaphore release on I2M completion is not supported!");
 
-        if (state.launchDma.layout == RegisterState::DmaDstMemoryLayout::Pitch && state.lineCount == 1) {
-            Logger::Debug("range: 0x{:X} -> 0x{:X}", u64{state.offsetOut}, u64{state.offsetOut} + buffer.size() * 0x4);
+        Logger::Debug("range: 0x{:X} -> 0x{:X}", u64{state.offsetOut}, u64{state.offsetOut} + buffer.size() * 0x4);
+        if (state.launchDma.layout == RegisterState::DmaDstMemoryLayout::Pitch) {
             channelCtx.channelSequenceNumber++;
-            interconnect.Upload(u64{state.offsetOut}, span{buffer});
+
+            u32 pitchOffset{};
+            auto srcBuffer{span{buffer}.cast<u8>()};
+            for (u32 line{}; line < state.lineCount; ++line, pitchOffset += state.pitchOut)
+                interconnect.Upload(u64{state.offsetOut + pitchOffset}, srcBuffer.subspan(state.lineLengthIn * line, state.lineLengthIn));
+
         } else {
             channelCtx.executor.Submit();
-            Logger::Warn("Non-linear I2M uploads are not supported!");
+
+            gpu::texture::Dimensions srcDimensions{state.lineLengthIn, state.lineCount, state.dstDepth};
+
+            gpu::texture::Dimensions dstDimensions{state.dstWidth, state.dstHeight, state.dstDepth};
+            size_t dstSize{GetBlockLinearLayerSize(dstDimensions, 1, 1, 1, 1 << (u8)state.dstBlockSize.height, 1 << (u8)state.dstBlockSize.depth)};
+
+            auto dstMappings{channelCtx.asCtx->gmmu.TranslateRange(state.offsetOut, dstSize)};
+
+            auto dst{dstMappings.front()};
+            [[maybe_unused]] std::vector<u8> tempBuffer;
+
+            if (dstMappings.size() != 1) {
+                // We create a temporary buffer to hold the blockLinear texture if mappings are split
+                // NOTE: We don't reserve memory here since such copies on this engine are rarely used
+                tempBuffer.resize(dstSize);
+                dst = span{tempBuffer};
+            }
+
+            // The I2M engine only supports a formatBpb of 1
+            if ((srcDimensions.width != dstDimensions.width) || (srcDimensions.height != dstDimensions.height))
```

Comment:
I think this might be better split out into a lambda, then the multiple mappings paths can be on either side of an if else rather having an if, <common code>, repeated if, which I find to be a bit awkward